lib/record_store/zone.rb

```diff
@@ -56,22 +56,49 @@ def yaml_files
     validate :validate_provider_can_handle_zone_records
 
     def self.from_yaml_definition(name, definition)
+      definition['records'] ||= []
+      Dir["#{RecordStore.zones_path}/#{name}/*__*.yml"].each do |record_file|
+        fqdn, type = File.basename(record_file, '.yml').split('__')
+        records_definition = Array.wrap(YAML.load_file(record_file))
+        records_definition.each do |record_definition|
+          record_definition[:fqdn] = fqdn
+          record_definition[:type] = type
+        end
+        definition['records'] += records_definition
+      end
       new(name, definition.deep_symbolize_keys)
     end
 
     def self.download(name, provider_name)
       dns = new(name, config: {provider: provider_name}).provider
-      current_records = dns.retrieve_current_records
-
-      File.write("#{RecordStore.zones_path}/#{name}.yml", {
-        name => {
-          config: {
-            provider: provider_name,
-            ignore_patterns: [{type: "NS", fqdn: "#{name}."}],
-          },
-          records: current_records.map(&:to_hash).sort_by! {|r| [r.fetch(:fqdn), r.fetch(:type), r[:nsdname] || r[:address]]}
-        }
-      }.deep_stringify_keys.to_yaml.gsub("---\n", ''))
+      current_records = dns.retrieve_current_records.map(&:to_hash)
+      write(name, records: current_records, config: {
+        provider: provider_name,
+        ignore_patterns: [{type: "NS", fqdn: "#{name}."}],
+      })
+    end
+
+    def self.write(name, config:, records:)
+      name = name.remove(/\.\z/)
+      zone_file = "#{RecordStore.zones_path}/#{name}.yml"
+      dir = "#{RecordStore.zones_path}/#{name}"
+      Dir["#{dir}/*"].each { |file| File.unlink(file) }
+      begin
+        Dir.mkdir(dir)
+      rescue Errno::EEXIST
+      end
+      zone = { name => { config: config } }
+      File.write(zone_file, zone.deep_stringify_keys.to_yaml.gsub("---\n", ''))
+      records.group_by { |record| [record.fetch(:fqdn), record.fetch(:type)] }.each do |(fqdn, type), grouped_records|
+        grouped_records.sort_by! {|record| record[:nsdname] || record[:address] }
+        grouped_records.each do |record|
+          record.delete(:fqdn)
+          record.delete(:type)
+          record.deep_stringify_keys!
+        end
+        grouped_records = grouped_records.first if grouped_records.size == 1
+        File.write("#{dir}/#{fqdn.remove(/\.\z/)}__#{type}.yml", grouped_records.to_yaml.gsub("---\n", ''))
```

Comment:
As implied by an earlier comment, `Zone#write`/`Zone.write` should be able to save the zone in both formats (single yaml file and directory of records).